File: runtime/trailer.go
package main

// Payload trailer parsing. Must mirror src/axe/trailer.py exactly:
//
//	[stub][payload zip][u64 payload off][u64 payload len][u32 version]["AXEBIN01"]
//
// The payload zip holds config.json, wheels/, uv/<archive>, python/<archive> —
// everything needed to bootstrap with zero network access.

import (
	"archive/zip"
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"os"
	"slices"
	"strings"
)

const (
	trailerMagic         = "AXEBIN01"
	trailerSize          = 8 + 8 + 4 + 8
	trailerFormatVersion = 2
)

var errNoTrailer = errors.New("no axe payload found in executable (was this binary built with `axe build`?)")

// payload gives access to the zip archive appended to the executable. The
// underlying file stays open for the process lifetime.
type payload struct {
	zip *zip.Reader
}

func openPayload(exePath string) (*payload, error) {
	f, err := os.Open(exePath)
	if err != nil {
		return nil, fmt.Errorf("cannot open own executable: %w", err)
	}
	info, err := f.Stat()
	if err != nil {
		return nil, err
	}
	if info.Size() < trailerSize {
		return nil, errNoTrailer
	}
	buf := make([]byte, trailerSize)
	if _, err := f.ReadAt(buf, info.Size()-trailerSize); err != nil {
		return nil, err
	}
	if string(buf[20:]) != trailerMagic {
		return nil, errNoTrailer
	}
	if version := binary.LittleEndian.Uint32(buf[16:20]); version != trailerFormatVersion {
		return nil, fmt.Errorf("unsupported payload format version %d (runtime supports %d)", version, trailerFormatVersion)
	}
	offset := binary.LittleEndian.Uint64(buf[0:8])
	length := binary.LittleEndian.Uint64(buf[8:16])
	if offset+length > uint64(info.Size())-trailerSize {
		return nil, errors.New("corrupt trailer: payload extends past end of file")
	}
	zr, err := zip.NewReader(io.NewSectionReader(f, int64(offset), int64(length)), int64(length))
	if err != nil {
		return nil, fmt.Errorf("corrupt payload archive: %w", err)
	}
	return &payload{zip: zr}, nil
}

func (p *payload) open(name string) (io.ReadCloser, error) {
	f, err := p.zip.Open(name)
	if err != nil {
		return nil, fmt.Errorf("payload is missing %s: %w", name, err)
	}
	return f, nil
}

func (p *payload) readAll(name string) ([]byte, error) {
	f, err := p.open(name)
	if err != nil {
		return nil, err
	}
	defer f.Close()
	return io.ReadAll(f)
}

func (p *payload) configBytes() ([]byte, error) {
	return p.readAll("config.json")
}

// wheelNames lists the payload's wheels/ entries, sorted so extraction order
// never depends on how the payload was assembled.
func (p *payload) wheelNames() []string {
	var names []string
	for _, f := range p.zip.File {
		if strings.HasPrefix(f.Name, "wheels/") && !strings.HasSuffix(f.Name, "/") {
			names = append(names, f.Name)
		}
	}
	slices.Sort(names)
	return names
}

File: src/axe/trailer.py
"""Binary payload trailer: how a stub finds the payload appended to it.

Layout of a built binary (format version 2):

    [stub executable bytes]
    [payload: a single ZIP archive]
    [trailer, 28 bytes:
        payload offset u64 LE
        payload length u64 LE
        format version u32 LE
        magic          "AXEBIN01"]

The payload zip contains everything the first run needs, so no network access
is ever required on the user's machine:

    config.json                 runtime configuration
    wheels/*.whl                the app wheel + all dependency wheels
    uv/<uv release artifact>    uv-<triple>.tar.gz (or .zip on Windows)
    python/<pbs artifact>       python-build-standalone install_only tarball

The Go runtime (runtime/trailer.go) reads this backwards from EOF; both sides
must agree exactly.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass

from . import TRAILER_FORMAT_VERSION

MAGIC = b"AXEBIN01"
TRAILER_STRUCT = struct.Struct("<QQI8s")
TRAILER_SIZE = TRAILER_STRUCT.size  # 28


class TrailerError(Exception):
    pass


@dataclass
class Payload:
    data: bytes  # the payload zip


def pack(stub: bytes, payload: bytes) -> bytes:
    trailer = TRAILER_STRUCT.pack(
        len(stub),
        len(payload),
        TRAILER_FORMAT_VERSION,
        MAGIC,
    )
    return stub + payload + trailer


def unpack(binary: bytes) -> Payload:
    if len(binary) < TRAILER_SIZE:
        raise TrailerError("binary too small to contain a trailer")
    offset, length, version, magic = TRAILER_STRUCT.unpack(binary[-TRAILER_SIZE:])
    if magic != MAGIC:
        raise TrailerError("no axe payload trailer found (bad magic)")
    if version != TRAILER_FORMAT_VERSION:
        raise TrailerError(f"unsupported trailer format version {version}")
    if offset + length > len(binary) - TRAILER_SIZE:
        raise TrailerError("corrupt trailer: payload extends past end of file")
    return Payload(data=binary[offset : offset + length])
